File: cpp/testsuite/OfflineExamples/RaytracingOfflineExample/src/RaytracerParallelExecutor.cpp
#include "RaytracerParallelExecutor.h"

#include "vsdk/toolkit/environment/material/RendererConfiguration.h"
#include "vsdk/toolkit/environment/scene/SimpleSceneSnapshot.h"
#include "vsdk/toolkit/gui/feedback/ProgressMonitor.h"
#include "vsdk/toolkit/gui/feedback/parallel/ParallelProgressMonitorConsumer.h"
#include "vsdk/toolkit/gui/feedback/parallel/ParallelProgressMonitorEvent.h"
#include "vsdk/toolkit/gui/feedback/parallel/ParallelProgressMonitorProducer.h"
#include "java/util/concurrent/Callable.h"
#include "java/util/concurrent/ConcurrentLinkedQueue.h"
#include "java/util/concurrent/ExecutorService.h"
#include "java/util/concurrent/Executors.h"
#include "java/util/concurrent/Future.h"
#include "java/util/concurrent/Void.h"
#include "vsdk/toolkit/media/RGBImageUncompressed.h"
#include "vsdk/toolkit/render/SimpleRaytracer.h"
#include "vsdk/toolkit/render/Tile.h"
#include "vsdk/toolkit/render/TileGenerationStrategy.h"
#include "vsdk/toolkit/render/TileGenerator.h"

#include <pthread.h>
#include <unistd.h>
#include <cstdio>
#include "vsdk/toolkit/common/VSDKFatalException.h"
#include "vsdk/toolkit/common/logging/Logger.h"
#include "java/util/ArrayList.txx"

struct ConsumerThreadData {
    ParallelProgressMonitorConsumer* consumer;
};

void* RaytracerParallelExecutor::progressConsumerMain(void* arg)
{
    ConsumerThreadData* data = reinterpret_cast<ConsumerThreadData*>(arg);
    data->consumer->run();
    return 0;
}

class TileWorker : public java::Callable<java::Void> {
private:
    java::ConcurrentLinkedQueue<Tile>* pendingTiles;
    RGBImageUncompressed* resultingImage;
    const RendererConfiguration* rendererConfiguration;
    SimpleSceneSnapshot* sceneSnapshot;
    ProgressMonitor* progressReporter;

public:
    TileWorker(java::ConcurrentLinkedQueue<Tile>* pendingTiles,
               RGBImageUncompressed* resultingImage,
               const RendererConfiguration* rendererConfiguration,
               SimpleSceneSnapshot* sceneSnapshot,
               ProgressMonitor* progressReporter)
        : pendingTiles(pendingTiles),
          resultingImage(resultingImage),
          rendererConfiguration(rendererConfiguration),
          sceneSnapshot(sceneSnapshot),
          progressReporter(progressReporter)
    {
    }

    virtual java::Void call() override
    {
        Tile tile(resultingImage, 0, 0, resultingImage->getXSize(), resultingImage->getYSize());
        SimpleRaytracer raytracer;

        while ( pendingTiles->poll(&tile) ) {
            raytracer.execute(resultingImage,
                             rendererConfiguration,
                             sceneSnapshot,
                             progressReporter,
                             0,
                             tile.getX0(),
                             tile.getY0(),
                             tile.getX1(),
                             tile.getY1());
        }
        return java::Void();
    }
};

long long RaytracerParallelExecutor::calculateTotalProgressElements(const java::ArrayList<Tile>& generatedTiles)
{
    long long total = 0;
    for (long int i = 0; i < generatedTiles.size(); i++) {
        total += generatedTiles.get(i).getDy();
    }
    return total;
}

void RaytracerParallelExecutor::run(SimpleRaytracer*,
                                    RGBImageUncompressed* resultingImage,
                                    const RendererConfiguration* rendererConfiguration,
                                    SimpleSceneSnapshot* sceneSnapshot,
                                    ProgressMonitor*)
{
#ifdef VITRAL_WITH_POSIX_THREADS
    long cpuCount = sysconf(_SC_NPROCESSORS_ONLN);
    int numberOfThreads = cpuCount > 0 ? (int)cpuCount : 1;
#else
    int numberOfThreads = 1;
#endif

    TileGenerator tileGenerator(TileGenerationStrategy::LINEAR,
                                resultingImage,
                                resultingImage->getXSize(),
                                resultingImage->getYSize(),
                                numberOfThreads);
    printf("Starting parallel raytracing with %d threads.\n", numberOfThreads);

    java::ArrayList<Tile> generatedTiles = tileGenerator.getTiles();
    java::ConcurrentLinkedQueue<Tile> pendingTiles(generatedTiles);
    java::ExecutorService* executorService =
        java::Executors::newFixedThreadPool(numberOfThreads);

    java::ConcurrentLinkedQueue<ParallelProgressMonitorEvent> progressEvents;
    ParallelProgressMonitorProducer producer(&progressEvents);
    ParallelProgressMonitorConsumer consumer(&progressEvents);

    producer.init(calculateTotalProgressElements(generatedTiles));

#ifdef VITRAL_WITH_POSIX_THREADS
    pthread_t consumerThread;
    ConsumerThreadData consumerData;
    consumerData.consumer = &consumer;
    pthread_create(&consumerThread, 0, &progressConsumerMain, &consumerData);
#else
    consumer.run();
#endif

    java::ArrayList<java::Future<java::Void> > futures;
    for (int i = 0; i < numberOfThreads; i++) {
        futures.add(executorService->submit(new TileWorker(
            &pendingTiles,
            resultingImage,
            rendererConfiguration,
            sceneSnapshot,
            &producer)));
    }

    for (long int i = 0; i < futures.size(); i++) {
        futures[i].get();
    }

    producer.finish();
    executorService->shutdownNow();
    delete executorService;

#ifdef VITRAL_WITH_POSIX_THREADS
    pthread_join(consumerThread, 0);
#endif

    if ( !pendingTiles.isEmpty() ) {
        Logger::reportMessage("RaytracerParallelExecutor", Logger::ERROR, "run", "Parallel raytracing finished with pending tiles");
        throw VSDKFatalException("Parallel raytracing finished with pending tiles");
    }
}

File: java/testsuite/OfflineExamples/RaytracingOfflineExample/src/RaytracerParallelExecutor.java
import java.util.ArrayList;
import java.util.List;
import java.util.concurrent.Callable;
import java.util.concurrent.ConcurrentLinkedQueue;
import java.util.concurrent.ExecutionException;
import java.util.concurrent.ExecutorService;
import java.util.concurrent.Executors;
import java.util.concurrent.Future;

import vsdk.toolkit.environment.material.RendererConfiguration;
import vsdk.toolkit.environment.scene.SimpleSceneSnapshot;
import vsdk.toolkit.gui.feedback.ProgressMonitor;
import vsdk.toolkit.gui.feedback.parallel.ParallelProgressMonitorConsumer;
import vsdk.toolkit.gui.feedback.parallel.ParallelProgressMonitorEvent;
import vsdk.toolkit.gui.feedback.parallel.ParallelProgressMonitorProducer;
import vsdk.toolkit.media.RGBImageUncompressed;
import vsdk.toolkit.render.raytracing.SimpleRaytracer;
import vsdk.toolkit.render.raytracing.RasterTileArea;
import vsdk.toolkit.render.raytracing.RasterTileGenerationStrategy;
import vsdk.toolkit.render.raytracing.RasterTileGenerator;

final class RaytracerParallelExecutor implements RaytracerExecutor {
    @Override
    public void run(SimpleRaytracer visualizationEngine,
                    RGBImageUncompressed resultingImage,
                    RendererConfiguration rendererConfiguration,
                    SimpleSceneSnapshot sceneSnapshot,
                    ProgressMonitor reporter)
    {
        int numberOfThreads = Runtime.getRuntime().availableProcessors();
        RasterTileGenerator tileGenerator = new RasterTileGenerator(
            RasterTileGenerationStrategy.LINEAR,
            resultingImage,
            resultingImage.getXSize(),
            resultingImage.getYSize(),
            numberOfThreads);
        System.out.println(
            "Starting parallel raytracing with " + numberOfThreads + " threads.");
        List<RasterTileArea> generatedTiles = tileGenerator.getTiles();
        ConcurrentLinkedQueue<RasterTileArea> pendingTiles =
            new ConcurrentLinkedQueue<>(generatedTiles);
        ExecutorService executorService = Executors.newFixedThreadPool(numberOfThreads);
        ConcurrentLinkedQueue<ParallelProgressMonitorEvent> progressEvents =
            new ConcurrentLinkedQueue<>();
        ParallelProgressMonitorProducer producer =
            new ParallelProgressMonitorProducer(progressEvents);
        ParallelProgressMonitorConsumer consumer =
            new ParallelProgressMonitorConsumer(progressEvents);
        Thread consumerThread = new Thread(consumer, "parallel-progress-monitor-consumer");
        producer.init(calculateTotalProgressElements(generatedTiles));
        consumerThread.start();

        try {
            List<Future<Void>> futures =
                startWorkers(
                    executorService,
                    numberOfThreads,
                    pendingTiles,
                    resultingImage,
                    rendererConfiguration,
                    sceneSnapshot,
                    producer);
            awaitWorkers(pendingTiles, futures);
        }
        catch ( InterruptedException e ) {
            Thread.currentThread().interrupt();
            throw new IllegalStateException("Parallel raytracing was interrupted", e);
        }
        catch ( ExecutionException e ) {
            throw new IllegalStateException("Parallel raytracing failed", e);
        }
        finally {
            producer.finish();
            executorService.shutdownNow();
            try {
                consumerThread.join();
            }
            catch ( InterruptedException e ) {
                Thread.currentThread().interrupt();
            }
        }
    }

    private long calculateTotalProgressElements(List<RasterTileArea> generatedTiles)
    {
        long totalElements = 0;

        for ( RasterTileArea tile : generatedTiles ) {
            totalElements += tile.getDy();
        }
        return totalElements;
    }

    private List<Future<Void>> startWorkers(
        ExecutorService executorService,
        int numberOfThreads,
        ConcurrentLinkedQueue<RasterTileArea> pendingTiles,
        RGBImageUncompressed resultingImage,
        RendererConfiguration rendererConfiguration,
        SimpleSceneSnapshot sceneSnapshot,
        ProgressMonitor progressReporter)
    {
        ArrayList<Future<Void>> futures =
            new ArrayList<>(numberOfThreads);

        for ( int i = 0; i < numberOfThreads; i++ ) {
            futures.add(executorService.submit(new TileWorker(
                pendingTiles,
                resultingImage,
                rendererConfiguration,
                sceneSnapshot,
                progressReporter)));
        }
        return futures;
    }

    private void awaitWorkers(
        ConcurrentLinkedQueue<RasterTileArea> pendingTiles,
        List<Future<Void>> futures)
        throws InterruptedException, ExecutionException
    {
        for ( Future<Void> future : futures ) {
            future.get();
        }
        if ( !pendingTiles.isEmpty() ) {
            throw new IllegalStateException("Parallel raytracing finished with pending tiles");
        }
    }

    private record TileWorker(
        ConcurrentLinkedQueue<RasterTileArea> pendingTiles,
        RGBImageUncompressed resultingImage,
        RendererConfiguration rendererConfiguration,
        SimpleSceneSnapshot sceneSnapshot,
        ProgressMonitor progressReporter)
        implements Callable<Void> {

        @Override
        public Void call()
        {
            RasterTileArea tile;
            SimpleRaytracer raytracer = new SimpleRaytracer();

            while ( (tile = pendingTiles.poll()) != null ) {
                raytracer.execute(
                    resultingImage,
                    rendererConfiguration,
                    sceneSnapshot,
                    progressReporter,
                    null,
                    tile.getX0(),
                    tile.getY0(),
                    tile.getX1(),
                    tile.getY1());
            }

            return null;
        }
    }
}
